app/controllers/settings/ext_cert_controller.rb

```diff
@@ -0,0 +1,260 @@
+require "openssl"
+
+# Settings::Ext_Cert allows users to install their own SSL Certificates
+# and Private Keys for encrypted external communication
+# rubocop:disable Metrics/ClassLength
+class Settings::ExtCertController < SettingsController
+  VELUM_NAME = "Velum".freeze
+  KUBEAPI_NAME = "Kubernetes API".freeze
+  DEX_NAME = "Dex".freeze
+
+  def index
+    set_instance_variables
+  end
+
+  def create
+    key_cert_map_temp = key_cert_map
+    key_cert_map_temp.each_key do |i|
+      return false unless upload_validate(key_cert_map_temp[i])
+    end
+
+    cert_map = {
+      external_cert_velum_cert:   key_cert_map_temp[:velum][:cert][:cert_string],
+      external_cert_velum_key:    key_cert_map_temp[:velum][:key][:key_string],
+      external_cert_kubeapi_cert: key_cert_map_temp[:kubeapi][:cert][:cert_string],
+      external_cert_kubeapi_key:  key_cert_map_temp[:kubeapi][:key][:key_string],
+      external_cert_dex_cert:     key_cert_map_temp[:dex][:cert][:cert_string],
+      external_cert_dex_key:      key_cert_map_temp[:dex][:key][:key_string]
+    }
+    @errors = Pillar.apply cert_map
+    if @errors.empty?
+      redirect_to settings_ext_cert_index_path,
+        notice: "External Certificate settings successfully saved."
+      return
+    # :nocov:
+    # An error here would require a failure in connection to velum->salt
+    # or a corruption in mapping of values in the salt pillar
+    else
+      set_instance_variables
+      render action: :index, status: :unprocessable_entity
+      # :nocov:
+    end
+  end
+
+  private
+
+  def set_instance_variables
+    @velum_cert = cert_parse(Pillar.value(pillar: :external_cert_velum_cert))
+    @velum_key = key_parse(Pillar.value(pillar: :external_cert_velum_key))
+    @kubeapi_cert = cert_parse(Pillar.value(pillar: :external_cert_kubeapi_cert))
+    @kubeapi_key = key_parse(Pillar.value(pillar: :external_cert_kubeapi_key))
+    @dex_cert = cert_parse(Pillar.value(pillar: :external_cert_dex_cert))
+    @dex_key = key_parse(Pillar.value(pillar: :external_cert_dex_key))
+  end
+
+  def get_val_from_form(param)
+    if params.key?(:external_certificate) && params[:external_certificate].key?(param)
+      params[:external_certificate][param].read.strip
```

Comment:
I don't think the extra 'strip' belongs. The direct contents of the file should be used. The strip was there originally in the code this was copied from to handle paste issues. The same will not be true for a file upload.

Also, what happens if a zero line file is uploaded? Am curious.

Reply:
Fair enough.  .strip on an empty string is safe.